### DISCUSSION TITLE
Adding edtra EXIF data

### DISCUSSION THREAD

**Discussion by czimm74** [aaronwmorris/indi-allsky#704] [Ideas]
Would be nice to add more exif data into the image file when writing.  Similar to how the Label Image Template is.    That way you could store all the camera, lens, location, all the image label items in the data of the image.  Then can query that information from something like Darkroom or any other image editor.

I have just started looking through the code to see that the temp exif json file is I see during the image capture.

**Comment by aaronwmorris** [comment]
I can work on this soon.

This is one of the major complaints of the OpenCV library, it does not support writing EXIF data.  I need to rip out all of the OpenCV imread/imwrite functionality and replace it with PIL or imageio.

**Comment by czimm74** [comment]
Just things I think of when I am trying to bre... I mean improve things.

**Comment by StarGeezerPhil** [comment]
Another thing that'd be interesting to explore would be including EXIF projection, panorama, fov details to enable viewing of fisheye captures within a suitable 360 viewer - or allow software to flatten or otherwise process the images more easily.

I'm not sure of all the details, but know there are some meta/EXIF properties to support some of this.

It's something I was going to look at to do some local processing in PHP after the end of night script. I really should learn some Python and dig into your code Aaron 🤣

**Reply by aaronwmorris** [reply]
I am currently replacing all of the OpenCV imread/imwrite functions with Pillow in order to support EXIF.

The side benefit is Pillow appears to be faster by at least 50%.  Actions that perform a lot of image reads and writes like generating keograms and startrails, which can take 10-15 minutes, can save as much as 60s of run time.

**Comment by aaronwmorris** [comment]
Bleh... Building exif data [in python] is more difficult than I thought it would be.

**Comment by czimm74** [comment]
Been doing some looking into it myself.  No one seems to have a clean way of doing it easily.  I wonder if running it post process, query the data from the sqllite data would be easier.

**Reply by aaronwmorris** [reply]
I finally figured out the piexif module.  The module itself is effectively abandoned, but will still work for my purposes.

**Comment by aaronwmorris** [comment]
I have merged #710 which includes EXIF metadata for all JPEG images.  I have tried to include as much detail as possible, including geolocation GPS tags and copyright information (if defined).  I have not done anything with PNG metadata, as that does not appear to be well standardized.

Most of the image I/O has been replaced with Pillow, which improves image performance by 25-50%.  Generating Keograms and Star Trails should be about 5-10% faster.

**Comment by czimm74** [comment]
ok, will run an update and see what I can blow up ;)

**Reply by czimm74** [reply]
Ok, threw an error:

Collecting pip
  Downloading https://www.piwheels.org/simple/pip/pip-23.1.1-py3-none-any.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╺━━━━━━━ 1.7/2.1 MB 22.4 kB/s eta 0:00:19
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
    pip from https://www.piwheels.org/simple/pip/pip-23.1.1-py3-none-any.whl#sha256=53a1498381da8ec323312a2c6a506b8ef7fd54d033a635f8df8ab1e443fedf8a:
        Expected sha256 53a1498381da8ec323312a2c6a506b8ef7fd54d033a635f8df8ab1e443fedf8a
             Got        0523a4de25c44971c0dd3942c87965b09900f4810763700b8cdc4944bc5495d2

**Reply by aaronwmorris** [reply]
That's a problem with pip, I would just try again.

**Reply by czimm74** [reply]
Took me a reboot first.  But running the update now without issues. 

Not capturing images, here are the errors that came up in the log:

Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: 
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: TypeError: Cannot handle this data type: (1, 1, 3), <f4
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:     raise TypeError(msg) from e
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:   File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/PIL/Image.py", line 3092, in fromarray
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:     img_rgb = Image.fromarray(cv2.cvtColor(data, cv2.COLOR_BGR2RGB))
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:   File "/home/pi/indi-allsky/indi_allsky/image.py", line 1205, in write_img
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:     latest_file, new_filename = self.write_img(self.image_processor.image, i_ref, camera, jpeg_exif=jpeg_exif)
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:   File "/home/pi/indi-allsky/indi_allsky/image.py", line 556, in processImage
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:     self.processImage(i_dict)
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:   File "/home/pi/indi-allsky/indi_allsky/image.py", line 257, in saferun
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:     self.saferun()
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:   File "/home/pi/indi-allsky/indi_allsky/image.py", line 228, in run
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: Traceback (most recent call last):
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: 
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: The above exception was the direct cause of the following exception:
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: 
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: KeyError: ((1, 1, 3), '<f4')
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:     mode, rawmode = _fromarray_typemap[typekey]
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception:   File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/PIL/Image.py", line 3089, in fromarray
Apr 25 17:32:50 indi-allsky [ERROR] MainProcess allsky._startImageWorker() #768: Image worker exception: Traceback (most recent call last):

**Reply by aaronwmorris** [reply]
Do you have the logo mask enabled?

**Reply by czimm74** [reply]
I have a file in there for the logo.  Have not made any changes to that, will try taking the image out and see what happens.

**Reply by czimm74** [reply]
Ok, I took the logo mask image out.  And the image processing worked.  Will try re-uploading the image and enable again and see what happens

**Reply by aaronwmorris** [reply]
I just needed to narrow down where the error originated and I noticed I was not casting the logo data back to uint8.

**Reply by czimm74** [reply]
Only works is the logo overlay is blank, will check if exif data is being written soon

**Reply by aaronwmorris** [reply]
I have merged #716 with a fix for the logo mask.

**Reply by czimm74** [reply]
Ok, looks like it is writing exif data.   Will update and try your logo fix next

**Reply by czimm74** [reply]
Looks like you fixed the logo overlay.  No errors being thrown.

**Reply by aaronwmorris** [reply]
When I was using OpenCV to write images, it did not really care of the data was represented as integers or floats, it would automatically correct for it.  However, Pillow is really picky about the data being 8-bit unsigned ints.
